packages/config/src/utils/extensions/auto-install-extensions.ts

```diff
@@ -0,0 +1,138 @@
+        return {
+          slug: installableExt.slug,
+          error: {
+            code: 'FAILED_TO_FETCH_EXTENSION',
+            message: 'Failed to fetch extension host site url',
+          },
+        }
+      }
+
+      console.log(
+        `Installing extension "${installableExt.name ?? installableExt.slug}" on team "${
+          account.name
+        }" required by package(s): "${installableExt.meta.packages.join('",')}"`,
+      )
+
+      return installExtension({
+        accountId: accountId,
+        netlifyToken: opts.token,
+        slug: installableExt.slug,
+        hostSiteUrl: installableExt.hostSiteUrl,
+      })
+    }),
+  )
+
+  results.forEach(({ error, slug }) => {
+    if (error) {
+      console.warn(`Failed to install "${slug}" extension on team "${account.name}": ${error.message}`)
+      return
+    }
+    const ext = installableExtensions.find((ext) => ext.slug === slug)
+    console.log(`Installed${ext?.name ? ` "${ext.name}" ` : ` "${slug}" `}extension on team "${account.name}"`)
+  })
+}
+
+async function getExtensionsToInstall(opts: $TSFixMe, api: NetlifyAPI) {
+  if (!opts.accountId || !api.accessToken) {
+    // skip installing extensions if not logged in
+    return []
+  }
+
+  const accountId = opts.accountId
+  const siteId = opts.siteId
+
+  const getPackageJSON = async () => {
+    const packageJsonPath = `${opts.repositoryRoot}/package.json`
+    try {
+      const packageJson = await fs.readFile(packageJsonPath, 'utf8')
+      return JSON.parse(packageJson)
+    } catch (e) {
+      console.error(`[@netlify/config] Failed to read ${packageJsonPath}`, e)
+      return {}
+    }
+  }
+
+  const [autoInstallableExtensions, packageJson, installedExtensions] = await Promise.all([
+    getAutoInstallableExtensions(),
+    getPackageJSON(),
+    getInstalledExtensionsForSite({
+      accountId: accountId,
+      siteId: siteId,
+      netlifyToken: opts.token,
+    }),
+  ])
+
+  const autoInstallExtensions = autoInstallableExtensions.filter((extension) => {
+    return extension.packages.some((pkg) => {
+      return packageJson?.dependencies?.[pkg]
+    })
+  })
+
+  const extensionsToInstallMeta = autoInstallExtensions.filter((extension) => {
+    return !installedExtensions.find((installedExtension) => installedExtension.integrationSlug === extension.slug)
+  })
+
+  const extensionsToInstall = (
+    await Promise.all(
+      extensionsToInstallMeta.map(async (extMeta) => {
+        const extensionData = await getExtension({
+          accountId: accountId,
+          netlifyToken: opts.token,
+          slug: extMeta.slug,
+        })
+
+        return {
+          slug: extMeta.slug,
+          ...(extensionData ?? {}),
+          meta: extMeta,
+        }
+      }),
+    )
+  ).filter(Boolean)
+
+  return extensionsToInstall
+}
+
+type AutoInstallableExtensionMeta = {
+  slug: string
+  packages: string[]
+}
+export async function getAutoInstallableExtensions() {
```

Comment:
This is another API call that we'll run on every build, even if someone doesn't have any extensions, right? What happens if this fails? Is there any way we could piggyback off of any existing calls we're already making to Jigsaw?

Reply:
I reduced the additional request made per build down to just this one.

I also made a [PR to jigsaw](https://github.com/netlify/jigsaw/pull/566) to update this endpoint with cache control headers and added the `hostSiteUrl` field so we don't need to make another request for that now.